File: 07_n_factorial.php
<?php

function permutations($string) {
    $length = strlen($string);
    if ($length == 1) {
        return [$string];
    }
    $result = [];
    for ($i = 0; $i < $length; $i++) {
        $firstChar = $string[$i];
        $remainingChars = substr($string, 0, $i) . substr($string, $i+1);
        $subPermutations = permutations($remainingChars);
        foreach ($subPermutations as $perm) {
            $result[] = $firstChar . $perm;
        }
    }
    return $result;
}

// time : O(n!) | space : ...

File: 07_n_factorial.py
def permute(elements):
    if len(elements) <= 1:
        return [elements]
    else:
        result = []
        for i in range(len(elements)):
            m = elements[i]
            remaining_elements = elements[:i] + elements[i+1:]
            for p in permute(remaining_elements):
                result.append([m] + p)
        return result
    
# time : O(n!) | space : ...